storage/s3.go

```diff
@@ -212,6 +243,84 @@ func (s *S3) List(ctx context.Context, url *url.URL, _ bool) <-chan *Object {
 	return objCh
 }
 
+// listObjects is used for cloud services that does not support S3
+// ListObjectsV2 API. I'm looking at you GCS.
+func (s *S3) listObjects(ctx context.Context, url *url.URL) <-chan *Object {
```

Comment:
listObjects() and listObjectsV2() are almost identical. I think we should reuse the code by writing a simple helper function. I am not sure but we can do something like:
```
list(ctx context.Context, url *url.URL, isRecursive bool, fn func())
// it will take closure (either ListObjectsWithContext or ListObjectsV2PagesWithContext) and called according to storage type.

Reply:
agreed, though we still need a control for v1/v2 conditions. i don't have an elegant solution.

Reply:
I also could not find. Let's merge this one until finding an elegant solution.